### PR TITLE
Fixes some upgrader issues

### LINKED ISSUE #7060: Database version not detected correctly in the upgrade.php script (when run through CLI)
#### Description
I'm in the process of testing an upgrade for my SMF 2.0.18 site and ran in to this error when running the upgrade.php script through cli (`php upgrade.php`)

```
Fatal error: Uncaught ArgumentCountError: Too few arguments to function smf_db_query(), 1 passed in /var/www/forum/upgrade.php on line 2594 and at least 2 expected in /var/www/forum/Sources/Subs-Db-mysql.php:361
```

### Steps to reproduce
1. Load in a 2.0.18 database
2. Run the cli upgrade script: `php upgrade.php`

### Environment (complete as necessary)
- Version/Git revision: 2.1 RC4
- Database Type: Mysql 
- Database Version: 8.0.23
- PHP Version: 8.0.5

### Additional information/references

It appears the `$database_version` detection is missing an `eval` call.
At this point the version_check code is being injected into version_compare as is:

In upgrade.php
```php
        if (empty($database_version))
	{
		$database_version = $databases[$db_type]['version_check'];
		$where_field_support = $db_type == 'mysql' && version_compare('5.0', $database_version, '<=');
	}
```

If I change this to:
```php
        if (empty($database_version))
	{
		$database_version = eval($databases[$db_type]['version_check']);
		$where_field_support = $db_type == 'mysql' && version_compare('5.0', $database_version, '<=');
	}
```

The upgrade continues as expected.

### LINKED ISSUE #7070: MariaDB not compatible any longer.
#### Description
due to #7065 i can no longer upgrade SMF 2.1, it fails to detect a valid version of the mysql db.
### Steps to reproduce
1. run upgrade.php when you are using MariaDB, it errors out that your database version does not meet requirements.
2. 

### Environment (complete as necessary)
- Version/Git revision: upgraded at midnight EDT last night
- Database Type:
- Database Version: MariaDB 10.4.10
- PHP Version: 7.3.12
- wampserver

https://i.imgur.com/54iM4mS.png PHP ini screenshot, i dont know why it has an older version listed here. using wamp server.

### Additional information/references

### LINKED ISSUE #7075: Upgrader: Hangs on first step via browser
#### Description
Note this is different from #7073 - which is via CLI.  It is failing much earlier when run via browser.

Note this is also different from #7066 - it fails on backup.  This is running without the backup selected.

Bottom line is, today, the upgrader cannot run no matter which set of options is selected.

Lots of different recent changes appear to have had adverse effects...  

image

### Steps to reproduce
1. Upgrade from 2.0 to 2.1
2. No backups requested
3. Via browser, not CLI

### Environment (complete as necessary)
- Version/Git revision: Current
- Database Type:  Mysql
- Database Version:  5.7
- PHP Version:  7.4

### LINKED ISSUE #7078: Installer: same MySQL version detection as Upgrader
#### Description
The installer uses the same `mysqli_get_server_info()` MySQL version detection as the upgrader which causes issues in detecting the proper version number for MariaDB, so just adding this issue so we can fix the installer at the same time as the upgrader.

### Environment (complete as necessary)
- Version/Git revision: Github latest

### Additional information/references
Refer #7070 for previous discussion on mysqli_get_server_info() and MariaDB.

### PR DESCRIPTION
**Pull request by jdarwood007**
Fixes #7075 and #7066 by adding the needed variable into the headers
Fixes #7070 by using the select method instead.  If this introduces issues because we no longer check the client we can reintroduce that.
Fixes #7060 by addressing that we need to use the function rather than string for the db version.
Fixes #7078 by making the changes the same in the installer
This will conflict with #7072 if it merges first.  I changed the eval call to a closure instead.  Easier to read and Syntax highlighting works in editors.
I also added a sanity check to ensure the function exists.

### CODE REVIEW

**Review by sbulen — COMMENTED** [review]
Good point made over on #7078 - we should apply the same version check changes to the installer.

Code looks good, though I could not test #7070 without a MariaDB.  

I was able to confirm #7066 and #7075 are both addressed.  I was able to run a variety of upgrades, via CLI & via browser, with & without backups, from 1.1 & 2.0, on small DBs and a larger forum (a copy of my prod forum).  All went perfectly.

I needed to test this alongside #7076 since some of my test DBs have no PMs.  So they work well together.

**Comment by sbulen** [comment]
I have confirmed that this PR addresses #7060 ...

Thought to be honest, I'm not sure why yet...  🙄

That broken logic highlighted by #7060 is clearly not invoked under normal circumstances - it has never worked.

File: other/upgrade.php [hunk 1]
```diff
@@ -2541,7 +2548,7 @@ function checkChange(&$change)
 	// Attempt to find a database_version.
 	if (empty($database_version))
 	{
-		$database_version = $databases[$db_type]['version_check'];
+		$database_version = $databases[$db_type]['version_check']();
```
Comment (jdarwood007) [inline other/upgrade.php:2551]: @sbulen This change fixes #7060

**Comment by jdarwood007** [comment]
@sbulen Per #7078 i've synced up the installer.  I don't have a working PG instance to test on right now to confirm I didn't break it.
I did some cleanup and removed the UTF-8 code in the installer, since we are requiring UTF-8 in 2.1 anyways, there isn't an option to not use it.  Left the sanity checks though.

**Comment by jdarwood007** [comment]
FYI, I have MariaDB so I can confirm the MariaDB specific changes are working as expected.

**Comment by sbulen** [comment]
I hope to do another set of tests tonight.

I'd like to get this & #7076 in sooner rather than later.

I have a pg test environment & will do a 2.0=>2.1 upgrade test there, also.

**Comment by illori** [comment]
this does fix the MariaDB issue, but i dont know if related to this or something else i am getting this error when running upgrade.php

Notice: Undefined index: normalize in C:\wamp64\www\github2.1new\Sources\Subs.php on line 7598

Fatal error: Uncaught Error: Function name must be a string in C:\wamp64\www\github2.1new\Sources\Subs.php:7598
Stack trace:
#0 C:\wamp64\www\github2.1new\Sources\Subs.php(5769): iri_to_url('https://www.sim...')
#1 C:\wamp64\www\github2.1new\upgrade.php(1840): fetch_web_data('https://www.sim...')
#2 C:\wamp64\www\github2.1new\upgrade.php(1757): cli_scheduled_fetchSMfiles()
#3 C:\wamp64\www\github2.1new\upgrade.php(2869): DeleteUpgrade()
#4 C:\wamp64\www\github2.1new\upgrade.php(3429): ConvertUtf8()
#5 C:\wamp64\www\github2.1new\upgrade.php(357): serialize_to_json()
#6 {main}
  thrown in C:\wamp64\www\github2.1new\Sources\Subs.php on line 7598

**Comment by jdarwood007** [comment]
As it relates to normalize and the iri_to_url, I would say that would be @Sesquipedalian who recently merged a fix for that.

**Comment by Sesquipedalian** [comment]
> As it relates to normalize and the iri_to_url, I would say that would be @Sesquipedalian who recently merged a fix for that.

Correct. Just do a quick merge of the latest from release-2.1 into this PR, @jdarwood007, and that error will go away.

**Comment by illori** [comment]
i had done an upgrade of my files to latest before i applied the changes here to upgrade.php, i still think there is an issue that needs to be fixed.

anyone else able to duplicate the error i got?

**Comment by sbulen** [comment]
I've been testing this alongside #7076 for the last couple of days and it looks good.  Lots of variants, including migrations from yabbse, smf 1.0, 1.1, 2.0 & 2.1.  Some CLI, some browser.  Some php7.4, some php8.  I've even done PG installs & a PG 2.0=>2.1 upgrade.  All good.

I had a couple random quirky issues that I don't think were related to this PR, so they were logged separately.

I did not encounter the issues Illori encountered above.  Not sure why, but I haven't...  

I think this looks good & that #7076 & #7077 should go in asap so folks can test upgrades.

**Review by sbulen — APPROVED** [review]

**Comment by illori** [comment]
i ran my test via cli on windows if that makes any difference.

**Comment by jdarwood007** [comment]
@Sesquipedalian Can you merge this?  Its confirmed working, but I do try to avoid merging my own PRs.

**Pull request merged**